src/datadog/base64.cpp

```diff
@@ -1,109 +1,100 @@
-    auto c3 = k_base64_table[static_cast<size_t>(in[i++])];
+    uint32_t c0 = k_base64_table[static_cast<size_t>(input[i++])];
+    uint32_t c1 = k_base64_table[static_cast<size_t>(input[i++])];
+    uint32_t c2 = k_base64_table[static_cast<size_t>(input[i++])];
+    uint32_t c3 = k_base64_table[static_cast<size_t>(input[i++])];
 
-    if (c0 == SENTINEL_VALUE || c1 == SENTINEL_VALUE || c2 == SENTINEL_VALUE ||
-        c3 == SENTINEL_VALUE) {
+    if (c0 == k_sentinel || c1 == k_sentinel || c2 == k_sentinel ||
+        c3 == k_sentinel) {
       return "";
     }
 
-    decoder.buffer =
-        static_cast<uint32_t>(0) | static_cast<uint32_t>(c0) << 26 |
-        static_cast<uint32_t>(c1) << 20 | static_cast<uint32_t>(c2) << 14 |
-        static_cast<uint32_t>(c3) << 8;
+    decoder.buffer = 0 | c0 << (32 - 6 * 1) | c1 << (32 - 6 * 2) |
+                     c2 << (32 - 6 * 3) | c3 << (32 - 6 * 4);
 
-    // NOTE(@dmehala): It might seem confusion to read those bytes in reverse
+    // NOTE(@dmehala): It might seem confusing to read those bytes input reverse
     // order. It is related to the architecture endianess. For now the set of
     // architecture we support (x86_64 and arm64) are all little-endian.
-    out.push_back(decoder.bytes[3]);
-    out.push_back(decoder.bytes[2]);
-    out.push_back(decoder.bytes[1]);
+    // TODO(@dgoffredo): I'd prefer an endian-agnostic implementation.
+    // nginx-datadog targets x86_64 and arm64 in its binary releases, but
+    // dd-trace-cpp targets any standard C++17 compiler.
+    output.push_back(decoder.bytes[3]);
+    output.push_back(decoder.bytes[2]);
+    output.push_back(decoder.bytes[1]);
   }
 
-  if ((in_size - i) < 4) return "";  // not padded input is not supported
+  // If padding is missing, return the empty string in lieu of an Error.
+  if ((in_size - i) < 4) return "";
 
-  auto c0 = k_base64_table[static_cast<size_t>(in[i++])];
-  auto c1 = k_base64_table[static_cast<size_t>(in[i++])];
-  auto c2 = k_base64_table[static_cast<size_t>(in[i++])];
-  auto c3 = k_base64_table[static_cast<size_t>(in[i++])];
+  uint32_t c0 = k_base64_table[static_cast<size_t>(input[i++])];
+  uint32_t c1 = k_base64_table[static_cast<size_t>(input[i++])];
+  uint32_t c2 = k_base64_table[static_cast<size_t>(input[i++])];
+  uint32_t c3 = k_base64_table[static_cast<size_t>(input[i++])];
 
-  if (c0 == SENTINEL_VALUE || c1 == SENTINEL_VALUE || c2 == SENTINEL_VALUE ||
-      c3 == SENTINEL_VALUE) {
+  if (c0 == k_sentinel || c1 == k_sentinel || c2 == k_sentinel ||
+      c3 == k_sentinel) {
     return "";
   }
 
-  decoder.buffer = static_cast<uint32_t>(0) | static_cast<uint32_t>(c0) << 26 |
-                   static_cast<uint32_t>(c1) << 20 |
-                   static_cast<uint32_t>(c2) << 14 |
-                   static_cast<uint32_t>(c3) << 8;
+  decoder.buffer = 0 | c0 << (32 - 6 * 1) | c1 << (32 - 6 * 2) |
+                   c2 << (32 - 6 * 3) | c3 << (32 - 6 * 4);
```

Comment:
I don't find all the arithmetic particularly helpful, it does the opposite of what intended IMO. Could not make that change?